state-commitments/history/history_commitment.go

```diff
@@ -0,0 +1,623 @@
+	lvLen := uint64(len(leaves))
+	if lvLen == 0 {
+		return nil, nil, errors.New("nil leaves")
+	}
+	if virtual == 0 {
+		return nil, nil, errors.New("virtual size cannot be zero")
+	}
+	if lvLen > virtual {
+		return nil, nil, fmt.Errorf("num leaves %d should be <= virtual %d", lvLen, virtual)
+	}
+	if index+1 > virtual {
+		return nil, nil, fmt.Errorf("index %d + 1 should be <= virtual %d", index, virtual)
+	}
+	logVirtual := uint(math.Log2Floor(virtual) + 1)
+	if err = h.populateFillers(&leaves[lvLen-1], logVirtual); err != nil {
+		return nil, nil, err
+	}
+
+	if index+1 > lvLen {
+		prefix, err = h.subtreeExpansion(leaves, index+1, 0, false)
+	} else {
+		prefix, err = h.subtreeExpansion(leaves[:index+1], index+1, 0, false)
+	}
+	if err != nil {
+		return nil, nil, err
+	}
+	tail, err = h.proof(index, leaves, virtual, 0)
+	return
+}
+
+// populateFillers returns a slice built recursively as
+// ret[0] = the passed in leaf
+// ret[i+1] = Hash(ret[i] + ret[i])
+//
+// Allocates n hashes
+// Computes n-1 hashes
+// Copies 1 hash
+func (h *historyCommitter) populateFillers(leaf *common.Hash, n uint) error {
+	if leaf == nil {
+		return errors.New("nil leaf pointer")
+	}
+	h.fillers = make([]common.Hash, n)
+	copy(h.fillers[0][:], (*leaf)[:])
+	for i := uint(1); i < n; i++ {
+		h.hashInto(&h.fillers[i], &h.fillers[i-1], &h.fillers[i-1])
+	}
+	return nil
+}
+
+// lastLeafProofPositions returns the positions in a virtual merkle tree
+// of the sibling nodes that need to be hashed with the last leaf at each
+// layer to compute the root of the tree.
+func lastLeafProofPositions(virtual nonZero) []treePosition {
+	if virtual == 1 {
+		return []treePosition{}
+	}
+	limit := nextPowerOf2(uint64(virtual))
+	depth := math.Log2Floor(limit)
+	positions := make([]treePosition, depth)
+	idx := uint64(virtual) - 1
+	for l := range positions {
+		positions[l] = sibling(idx, uint64(l))
+		idx = parent(idx)
+	}
+	return positions
+}
+
+// sibling returns the position of the sibling of the node at the given layer
+func sibling(index, layer uint64) treePosition {
+	return treePosition{layer: layer, index: index ^ 1}
+}
+
+// parent returns the index of the parent of the node in the next higher layer
+func parent(index uint64) uint64 {
+	return index >> 1
+}
+
+func nextPowerOf2(n uint64) uint64 {
+	if n == 0 {
+		return 1
+	}
+	n--         // Decrement n to handle the case where n is a power of 2
+	n |= n >> 1 // Propagate the highest bit set
+	n |= n >> 2
+	n |= n >> 4
+	n |= n >> 8
+	n |= n >> 16
+	n |= n >> 32
+	return n + 1 // Increment n to get the next power of 2
+}
+
+func trimTrailingZeroHashes(hashes []common.Hash) []common.Hash {
+	// Start from the end of the slice
+	for i := len(hashes) - 1; i >= 0; i-- {
+		if hashes[i] != emptyHash {
+			return hashes[:i+1]
+		}
+	}
+	// If all elements are zero, return an empty slice
+	return []common.Hash{}
+}
+
+func trimZeroes(hashes []common.Hash) []common.Hash {
```

Comment:
nit: I would call filterZeroes instead of trim Zeroes.

Reply:
Went with filterEmptyHashes, but I agree this is clearly a filter.